client-hc/src/main/java/com/graphhopper/api/GraphHopperGeocoding.java

```diff
@@ -0,0 +1,133 @@
+import com.graphhopper.util.exceptions.*;
+import com.graphhopper.util.shapes.GHPoint;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.ResponseBody;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static com.graphhopper.util.Helper.round6;
+import static com.graphhopper.util.Helper.toLowerCase;
+
+/**
+ * Client implementation for the GraphHopper Directions API Geocoding
+ *
+ * @author Robin Boldt
+ */
+public class GraphHopperGeocoding {
+
+    private final ObjectMapper objectMapper;
+    private OkHttpClient downloader;
+    private String routeServiceUrl;
+    private String key = "";
+
+    private final long DEFAULT_TIMEOUT = 5000;
+
+    public GraphHopperGeocoding() {
+        this("https://graphhopper.com/api/1/geocode");
+    }
+
+    public GraphHopperGeocoding(String serviceUrl) {
+        this.routeServiceUrl = serviceUrl;
+        downloader = new OkHttpClient.Builder().
+                connectTimeout(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS).
+                readTimeout(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS).
+                build();
+
+        objectMapper = new ObjectMapper();
+    }
+
+    public GraphHopperGeocoding setDownloader(OkHttpClient downloader) {
+        this.downloader = downloader;
+        return this;
+    }
+
+    public void setKey(String key) {
+        this.key = key;
+    }
+
+    public OkHttpClient getDownloader() {
+        return downloader;
+    }
+
+    private OkHttpClient getClientForRequest(GHGeocodingRequest request) {
+        OkHttpClient client = this.downloader;
+        if (request.hasTimeout()) {
+            long timeout = request.getTimeout();
+            client = client.newBuilder()
+                    .connectTimeout(timeout, TimeUnit.MILLISECONDS)
+                    .readTimeout(timeout, TimeUnit.MILLISECONDS)
+                    .build();
+        }
+
+        return client;
+    }
+
+    public GHGeocodingResponse geocode(GHGeocodingRequest request) {
+        String url = buildUrl(request);
+
+        try {
+            Request okRequest = new Request.Builder().url(url).build();
+            ResponseBody rspBody = getClientForRequest(request).newCall(okRequest).execute().body();
+            return objectMapper.readValue(rspBody.bytes(), GHGeocodingResponse.class);
+        } catch (Exception ex) {
+            throw new RuntimeException("Problem performing geocoding for " + url + ": " + ex.getMessage(), ex);
+        }
+    }
+
+    private String buildUrl(GHGeocodingRequest request) {
+        String url = routeServiceUrl + "?";
+
+        if (request.isReverse()) {
+            url += "reverse=true";
+            url += "&point=" + request.getLat() + "," + request.getLon();
+        } else {
+            url += "reverse=false";
+            url += "&q=" + request.getQuery();
+            if (!Double.isNaN(request.getLat()))
+                url += "&point=" + request.getLat() + "," + request.getLon();
+        }
```

Comment:
I would help here the user a bit: if reverse than throw an exception if the lat or lon is invalid. And if non-reverse throw exc if query is empty. Also adding the point can be moved outside of the if+else IMO to have just one line

Reply:
Thanks, good idea!